File: dcr/policy.py
"""Runtime decision policy — what to do with each piece of context.

The wiki's decision table, made executable. The hard part it leaves open is
`sufficient(L, q)`: there is no clean definition of the cheapest level that can
answer a query. This implements the practical proxy the spec recommends —
route by query type, then *measure* escalation and let the escalation rate be
the telemetry that catches a bad router — plus a de-escalation rule so a node
that keeps being read at L0 without yielding anything new collapses into a
cached fact.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from .ladder import L0, L1, L2, L3
from .nodes import CALCULATION, CONSTRAINT, DECISION, EVIDENCE, GOAL, OPEN_QUESTION, Node

QUOTE_EXACT = "quote_exact"
VALUE_LOOKUP = "value_lookup"
JUSTIFY = "justify"
RECOMPUTE = "recompute"
SUMMARIZE = "summarize"
OPEN = "open"

_PATTERNS = (
    (QUOTE_EXACT, re.compile(
        r"\b(quote|verbatim|exact(ly)?|word[- ]for[- ]word|literal|raw text|"
        r"exact wording|full text|transcript|as written|stack trace|"
        r"exact error|error message|log line)\b", re.I)),
    (RECOMPUTE, re.compile(
        # Deliberately narrow: a bare "how many" is a lookup far more often
        # than it is a computation, and mis-routing a lookup to L3 wastes an
        # execution and returns nothing.
        r"\b(recompute|recalculate|re-?run|compute|calculate|"
        r"(total|sum|average|cost) (of|for)|what.s the (cost|total|result)|"
        r"how much (does|did|would) .* cost)\b", re.I)),
    (JUSTIFY, re.compile(
        r"\b(why|justify|justification|explain|reason|because|on what basis|"
        r"how did (you|we)|what led to|audit|evidence for|prove|back(ing)? up)\b", re.I)),
    (SUMMARIZE, re.compile(r"\b(summar(y|ise|ize)|recap|overview|tl;?dr|gist)\b", re.I)),
    (VALUE_LOOKUP, re.compile(
        r"\b(what (is|was|are|were)|which|who|when|where|current|status of|value of|"
        r"how (much|many))\b", re.I)),
)

#: query type -> preferred admission level, cheapest sufficient first
ROUTING = {
    QUOTE_EXACT: (L0, L1, L2),
    VALUE_LOOKUP: (L2, L1, L0),
    JUSTIFY: (L2, L1, L0),
    RECOMPUTE: (L3, L2, L0),
    SUMMARIZE: (L1, L2, L0),
    OPEN: (L2, L1, L0),
}

#: kinds that are cheap, near-always relevant, and pinned into every turn
ALWAYS_ADMIT = (GOAL, CONSTRAINT)


@dataclass
class Decisions:
    """What the runtime decided about one node this turn — for telemetry
    and for `explain_plan()`."""

    node_id: str
    level: str | None
    admitted: bool
    prefetch: bool = False
    recompute: bool = False
    reason: str = ""


class DecisionPolicy:
    def __init__(
        self,
        deescalate_after: int = 3,
        escalation_confidence: float = 0.55,
    ) -> None:
        self.deescalate_after = deescalate_after
        self.escalation_confidence = escalation_confidence

    # -- routing -----------------------------------------------------------
    def classify(self, query: str) -> str:
        for qtype, pattern in _PATTERNS:
            if pattern.search(query or ""):
                return qtype
        return OPEN

    def preferred_level(self, node: Node, qtype: str, available: list[str]) -> str:
        """Cheapest level in the routing order that this node actually has."""
        for level in ROUTING.get(qtype, ROUTING[OPEN]):
            if level in available:
                if level == L0 and self._prefers_compact(node, qtype):
                    continue
                return level
        return available[0] if available else L2

    def _prefers_compact(self, node: Node, qtype: str) -> bool:
        # Never promote a settled, well-grounded claim to raw bytes just
        # because the router said L0 — level over-promotion is the failure
        # mode that destroys the cost model.
        return (
            qtype != QUOTE_EXACT
            and node.kind not in (EVIDENCE,)
            and node.confidence >= 0.9
        )

    def sufficient(self, level: str, qtype: str) -> bool:
        order = ROUTING.get(qtype, ROUTING[OPEN])
        return level in order[:2] if qtype != QUOTE_EXACT else level == L0

    def escalation_target(self, node: Node, current: str, available: list[str]) -> str | None:
        """Next richer level to try when the model reports insufficiency."""
        ladder_up = {L2: L1, L1: L0, L3: L0, L0: None}
        target = ladder_up.get(current)
        while target is not None and target not in available:
            target = ladder_up.get(target)
        return target

    def should_deescalate(self, node: Node) -> bool:
        """A node read at L0 repeatedly with nothing new extracted should be
        collapsed into a cached fact instead of re-quoted every turn."""
        l0_reads = node.meta.get("l0_reads", 0)
        yielded = node.meta.get("l0_yield", 0)
        return l0_reads >= self.deescalate_after and yielded == 0

    # -- the decision table ------------------------------------------------
    def decide(self, node: Node, qtype: str, available: list[str], *, in_plan: bool) -> Decisions:
        if node.status == "stale":
            return Decisions(node.node_id, None, False, recompute=True,
                             reason="stale — recompute or re-ground before admitting")
        if node.status == "superseded":
            return Decisions(node.node_id, None, False, reason="superseded")
        if not in_plan:
            return Decisions(node.node_id, None, False, reason="not referenced by current plan")
        level = self.preferred_level(node, qtype, available)
        reason = {
            L0: "query needs exact wording",
            L1: "referenced for gist only",
            L2: "compact state object is sufficient",
            L3: "answer is a function of known inputs",
        }[level]
        return Decisions(node.node_id, level, True, reason=reason)

    def pinned_kinds(self, qtype: str) -> tuple[str, ...]:
        if qtype == JUSTIFY:
            return ALWAYS_ADMIT + (DECISION,)
        if qtype == RECOMPUTE:
            return ALWAYS_ADMIT + (CALCULATION,)
        if qtype == OPEN:
            return ALWAYS_ADMIT + (OPEN_QUESTION,)
        return ALWAYS_ADMIT

File: src/policy.rs
//! Runtime decision policy — what to do with each piece of context.
//!
//! The wiki's decision table, made executable. The hard part it leaves open is
//! `sufficient(L, q)`: there is no clean definition of the cheapest level that
//! can answer a query. This implements the practical proxy the spec recommends
//! — route by query type, then *measure* escalation and let the escalation rate
//! be the telemetry that catches a bad router — plus a de-escalation rule so a
//! node that keeps being read at L0 without yielding anything new collapses
//! into a cached fact.

use std::collections::HashSet;

use crate::ladder::Level;
use crate::nodes::{Kind, Node, NodeIdx, Status};
use crate::planner::ActiveContext;
use crate::text::{contains_any, content_tokens};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum QueryType {
    QuoteExact,
    ValueLookup,
    Justify,
    Recompute,
    Summarize,
    Open,
}

impl QueryType {
    pub fn as_str(self) -> &'static str {
        match self {
            QueryType::QuoteExact => "quote_exact",
            QueryType::ValueLookup => "value_lookup",
            QueryType::Justify => "justify",
            QueryType::Recompute => "recompute",
            QueryType::Summarize => "summarize",
            QueryType::Open => "open",
        }
    }

    /// Preferred admission levels, cheapest sufficient first.
    pub fn routing(self) -> [Level; 3] {
        match self {
            QueryType::QuoteExact => [Level::L0, Level::L1, Level::L2],
            QueryType::ValueLookup => [Level::L2, Level::L1, Level::L0],
            QueryType::Justify => [Level::L2, Level::L1, Level::L0],
            QueryType::Recompute => [Level::L3, Level::L2, Level::L0],
            QueryType::Summarize => [Level::L1, Level::L2, Level::L0],
            QueryType::Open => [Level::L2, Level::L1, Level::L0],
        }
    }
}

const QUOTE_MARKERS: &[&str] = &[
    "quote",
    "verbatim",
    "exact",
    "exactly",
    "word for word",
    "word-for-word",
    "literal",
    "raw text",
    "exact wording",
    "full text",
    "transcript",
    "as written",
    "stack trace",
    "exact error",
    "error message",
    "log line",
];

const RECOMPUTE_MARKERS: &[&str] = &[
    "recompute",
    "recalculate",
    "rerun",
    "re-run",
    "compute",
    "calculate",
    "total of",
    "total for",
    "sum of",
    "sum for",
    "average of",
    "cost of",
    "cost for",
    "what's the cost",
    "what is the cost",
    "what's the total",
    "what is the total",
    "what's the result",
    "what is the result",
];

const JUSTIFY_MARKERS: &[&str] = &[
    "why",
    "justify",
    "justification",
    "explain",
    "reason",
    "because",
    "on what basis",
    "how did you",
    "how did we",
    "what led to",
    "audit",
    "evidence for",
    "prove",
    "back up",
    "backing up",
];

const SUMMARIZE_MARKERS: &[&str] = &[
    "summary",
    "summarise",
    "summarize",
    "recap",
    "overview",
    "tldr",
    "tl;dr",
    "gist",
];

const VALUE_MARKERS: &[&str] = &[
    "what is",
    "what was",
    "what are",
    "what were",
    "which",
    "who",
    "when",
    "where",
    "current",
    "status of",
    "value of",
    "how much",
    "how many",
];

/// Lexical overlap between a question and one rendered context line.
///
/// Shared deliberately. [`LocalReasoner`](crate::llm::LocalReasoner) scores
/// context lines with this to decide whether to emit `#ESCALATE`, and
/// [`DecisionPolicy::needs_escalation`] scores the *assembled window* with the
/// same function to reach the same decision without being told. That is the
/// whole basis for serving a harness that cannot signal: the escalation
/// judgement never depended on anything private to the model, only on the
/// query and the bytes the model was handed. If these two ever diverge, the
/// runtime-side check is no longer a stand-in for the model-side one, so there
/// is one function rather than two that agree today.
pub fn overlap(query: &str, body: &str) -> f32 {
    let q: HashSet<String> = content_tokens(query).into_iter().collect();
    if q.is_empty() {
        return 0.0;
    }
    // `server.ip = 10.0.9.7` should match "server ip", so split dotted keys.
    let expanded = format!("{} {}", body.replace('.', " "), body);
    let tokens: HashSet<String> = content_tokens(&expanded).into_iter().collect();
    let mut score = q.intersection(&tokens).count() as f32 / q.len() as f32;
    // A question is about the *subject*, so key matches count double.
    let head = body.split('=').next().unwrap_or("").replace('.', " ");
    let head_tokens: HashSet<String> = content_tokens(&head).into_iter().collect();
    let head_hits = q.intersection(&head_tokens).count();
    if head_hits > 0 {
        score += 0.5 * head_hits as f32 / q.len() as f32;
    }
    score
}

/// Kinds that are cheap, near-always relevant, and pinned into every turn.
pub const ALWAYS_ADMIT: [Kind; 2] = [Kind::Goal, Kind::Constraint];

/// What the runtime decided about one node this turn — for telemetry and for
/// `explain_plan()`.
#[derive(Debug, Clone)]
pub struct Decision {
    pub level: Option<Level>,
    pub admitted: bool,
    pub recompute: bool,
    pub reason: &'static str,
}

#[derive(Debug, Clone)]
pub struct DecisionPolicy {
    pub deescalate_after: u32,
    /// Below this a match is unusable. Mirrors `LocalReasoner::threshold`.
    pub answer_floor: f32,
    /// Between `answer_floor` and this, the best match is plausible but thin.
    /// Mirrors `LocalReasoner::escalate_below`.
    pub escalate_below: f32,
    /// Below `answer_floor` but at or above this, there is enough overlap that
    /// the raw span is worth a look rather than answering "I don't have that".
    pub escalate_from: f32,
}

impl Default for DecisionPolicy {
    fn default() -> Self {
        Self {
            deescalate_after: 3,
            answer_floor: 0.25,
            escalate_below: 0.6,
            escalate_from: 0.1,
        }
    }
}

impl DecisionPolicy {
    // -- routing -----------------------------------------------------------
    pub fn classify(&self, query: &str) -> QueryType {
        let q = query.to_lowercase();
        if contains_any(&q, QUOTE_MARKERS) {
            return QueryType::QuoteExact;
        }
        // Deliberately narrow: a bare "how many" is a lookup far more often
        // than it is a computation, and mis-routing a lookup to L3 wastes an
        // execution and returns nothing.
        if contains_any(&q, RECOMPUTE_MARKERS)
            || (contains_any(&q, &["how much"]) && contains_any(&q, &["cost"]))
        {
            return QueryType::Recompute;
        }
        if contains_any(&q, JUSTIFY_MARKERS) {
            return QueryType::Justify;
        }
        if contains_any(&q, SUMMARIZE_MARKERS) {
            return QueryType::Summarize;
        }
        if contains_any(&q, VALUE_MARKERS) {
            return QueryType::ValueLookup;
        }
        QueryType::Open
    }

    /// Cheapest level in the routing order that this node actually has.
    pub fn preferred_level(&self, node: &Node, qtype: QueryType, available: &[Level]) -> Level {
        for level in qtype.routing() {
            if available.contains(&level) {
                if level == Level::L0 && self.prefers_compact(node, qtype) {
                    continue;
                }
                return level;
            }
        }
        available.first().copied().unwrap_or(Level::L2)
    }

    /// Never promote a settled, well-grounded claim to raw bytes just because
    /// the router said L0 — level over-promotion is the failure mode that
    /// destroys the cost model.
    fn prefers_compact(&self, node: &Node, qtype: QueryType) -> bool {
        qtype != QueryType::QuoteExact && node.kind != Kind::Evidence && node.confidence >= 0.9
    }

    pub fn sufficient(&self, level: Level, qtype: QueryType) -> bool {
        if qtype == QueryType::QuoteExact {
            level == Level::L0
        } else {
            qtype.routing()[..2].contains(&level)
        }
    }

    /// Next richer level to try when the model reports insufficiency.
    pub fn escalation_target(&self, current: Level, available: &[Level]) -> Option<Level> {
        let mut target = match current {
            Level::L2 => Some(Level::L1),
            Level::L1 | Level::L3 => Some(Level::L0),
            Level::L0 => None,
        };
        while let Some(level) = target {
            if available.contains(&level) {
                return Some(level);
            }
            target = match level {
                Level::L2 => Some(Level::L1),
                Level::L1 | Level::L3 => Some(Level::L0),
                Level::L0 => None,
            };
        }
        None
    }

    /// A node read at L0 repeatedly with nothing new extracted should be
    /// collapsed into a cached fact instead of re-quoted every turn.
    pub fn should_deescalate(&self, node: &Node) -> bool {
        node.meta.l0_reads.get() >= self.deescalate_after && node.meta.l0_yield.get() == 0
    }

    // -- the decision table ------------------------------------------------
    pub fn decide(
        &self,
        node: &Node,
        qtype: QueryType,
        available: &[Level],
        in_plan: bool,
    ) -> Decision {
        if node.status == Status::Stale {
            return Decision {
                level: None,
                admitted: false,
                recompute: true,
                reason: "stale — recompute or re-ground before admitting",
            };
        }
        if node.status == Status::Superseded {
            return Decision {
                level: None,
                admitted: false,
                recompute: false,
                reason: "superseded",
            };
        }
        if !in_plan {
            return Decision {
                level: None,
                admitted: false,
                recompute: false,
                reason: "not referenced by current plan",
            };
        }
        let level = self.preferred_level(node, qtype, available);
        Decision {
            level: Some(level),
            admitted: true,
            recompute: false,
            reason: match level {
                Level::L0 => "query needs exact wording",
                Level::L1 => "referenced for gist only",
                Level::L2 => "compact state object is sufficient",
                Level::L3 => "answer is a function of known inputs",
            },
        }
    }

    /// The node this window is too compressed to answer from, if any —
    /// computed by the runtime instead of waiting for the model to say so.
    ///
    /// The documented poor fit was "reasoners that cannot signal": a harness
    /// with no way to emit `#ESCALATE` loses the mechanism that carries the
    /// exact-quote and buried-detail probes. But the model-side trigger is a
    /// function of the query and the rendered window and nothing else, so the
    /// runtime can evaluate it directly. See [`overlap`].
    ///
    /// This is a *proxy*, and the distinction matters for what may be claimed
    /// from it: a real model's sense of insufficiency is not this function.
    /// What it establishes is that the signal is reconstructible from what the
    /// runtime already holds, not that escalation no longer needs a model.
    pub fn needs_escalation(
        &self,
        query: &str,
        context: &ActiveContext,
        already: &HashSet<NodeIdx>,
    ) -> Option<NodeIdx> {
        let mut best: Option<(f32, &crate::planner::Admission)> = None;
        for entry in &context.entries {
            let score = overlap(query, &entry.rendered);
            if best.is_none_or(|(top, _)| score > top) {
                best = Some((score, entry));
            }
        }
        let (score, entry) = best?;
        if entry.level == Level::L0 || already.contains(&entry.node) {
            return None;
        }
        let wants_quote = self.classify(query) == QueryType::QuoteExact;
        // Below the floor with *some* overlap the detail may be in the raw
        // span; between the floor and `escalate_below` the compact form is
        // plausible but thin, which is the case escalation exists for.
        let thin = score < self.escalate_below && score >= self.answer_floor;
        let unusable = score < self.answer_floor && score >= self.escalate_from;
        if wants_quote || thin || unusable {
            Some(entry.node)
        } else {
            None
        }
    }

    pub fn pinned_kinds(&self, qtype: QueryType) -> Vec<Kind> {
        let mut kinds = ALWAYS_ADMIT.to_vec();
        match qtype {
            QueryType::Justify => kinds.push(Kind::Decision),
            QueryType::Recompute => kinds.push(Kind::Calculation),
            QueryType::Open => kinds.push(Kind::OpenQuestion),
            _ => {}
        }
        kinds
    }
}
